File: 0146-lru-cache/0146-lru-cache.cpp
class LRUCache {
public:
    struct ListNode {
        int key, val;
        ListNode* prev;
        ListNode* next;

        ListNode(int key, int value) : key(key), val(value), prev(NULL), next(NULL) {}
    };

    ListNode* head = new ListNode(0, 0);
    ListNode* tail = new ListNode(0, 0);
    int curr_cap = 0;
    int max_cap;
    map <int, ListNode*> hashmap;
    LRUCache(int capacity) {
        head->next = tail;
        tail->prev = head;
        max_cap = capacity;
    }

    int get(int key) {
        if (hashmap.find(key) == hashmap.end()) return -1;
        else {
            int ans = hashmap[key]->val;
            if (head->next != hashmap[key]) {
                ListNode* headnext = head->next;
                ListNode* curr = hashmap[key];
                ListNode* toinsert = new ListNode(curr->key, curr->val);
                ListNode* currprev = curr->prev;
                ListNode* currnext = curr->next;
                delete(curr);
                currprev->next = currnext;
                currnext->prev = currprev;
                head->next = toinsert;
                toinsert->prev = head;
                toinsert->next = headnext;
                headnext->prev = toinsert;
                hashmap[key] = toinsert;
            }
            return ans;
        }
    }

    void put(int key, int value) {
        if (curr_cap == max_cap && hashmap.find(key) == hashmap.end()) {
            ListNode* tailprev = tail->prev;
            ListNode* pre = tailprev->prev;
            pre->next = tail;
            tail->prev = pre;
            hashmap.erase(tailprev->key);
            delete(tailprev);
            curr_cap--;
        }
        if (hashmap.find(key) == hashmap.end()) curr_cap++;
        else {
            ListNode* curr = hashmap[key];
            hashmap.erase(key);
            ListNode* pre = curr->prev;
            ListNode* nxt = curr->next;
            delete(curr);
            pre->next = nxt;
            nxt->prev = pre;
        }
        ListNode* toinsert = new ListNode(key, value);
        ListNode* headnext = head->next;
        head->next = toinsert;
        toinsert->prev = head;
        toinsert->next = headnext;
        headnext->prev = toinsert;
        hashmap[key] = toinsert;

    }
};

File: 0146-lru-cache/0146-lru-cache.java
class Node{
    int key;
    int val;
    Node prev;
    Node next;
    Node(int key, int val){
        this.key = key;
        this.val = val;
    }
}

class LRUCache {

    int cap;
    Node head;
    Node tail;
    HashMap <Integer, Node> mp;

    public LRUCache(int capacity) {
        this.cap = capacity;
        mp = new HashMap<>();
        head = new Node(-1,-1);
        tail = new Node(-1, -1);
        head.next= tail;
        tail.prev = head;
    }
    
    public void deleteNode(Node node){
        Node pre = node.prev;
        Node nxt = node.next;
        pre.next = nxt;
        nxt.prev = pre;
    }

    public void insertAfterHead(Node node){
        Node tmp = head.next;
        head.next = node;
        node.prev = head;
        tmp.prev = node;
        node.next = tmp;
    }

    public int get(int key) {
        if(!mp.containsKey(key)){
            return -1;
        }
        Node node = mp.get(key);
        deleteNode(node);
        insertAfterHead(node);
        return node.val;
    }
    
    public void put(int key, int value) {
        if(mp.containsKey(key)){
            Node node = mp.get(key);
            node.val = value;
            deleteNode(node);
            insertAfterHead(node);
        }else{
            if(mp.size() == cap){
                Node lru = tail.prev;
                deleteNode(lru);
                mp.remove(lru.key);
            }
            Node node = new Node(key, value);
            insertAfterHead(node);
            mp.put(key, node);
        }
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */